### DISCUSSION TITLE
List based filtering

### DISCUSSION THREAD

**Discussion by ascii1010** [Azure/data-api-builder#1897] [Q&A] [answered]
Hi,
is list based filtering supported in DAB? 

e.g. filter: { id:{ in:[1,2,3,4] }} or filter: { id:{ nin:[1,2,3,4] }}

If not, is there an alternative approach that generates the same outcome?

**Comment by severussundar** [comment, marked as answer]
Hi @ascii1010, this can be achieved using the `or` operator.

For example: Let's say you have an entity titled `book` and items with ids :1, 2 and 3 needs to be selected.

Using the following filter clause will select only the items with ids equal to 1,2 or 3.

```graphql
{
  books(filter: { or: [{ id: { eq: 1 } }, { id: { eq: 2 } }, { id: { eq: 3 } }] }) {
    items {
      id
      ...
    }
  }
}
``` 

For reference, [here](https://learn.microsoft.com/azure/data-api-builder/graphql#filter) is the list of operators that are supported along with `filter`

**Reply by ascii1010** [reply]
@severussundar, thanks for your suggestion, it worked.

**Comment by alawal98** [comment]
Hi [severussundar](https://github.com/severussundar), is this currently possible for multiple fields? e.g. I want to get all the books that have ID 1-3 but I also want these books to be in the "Horror" or "Thriller" Genre.

Thanks.

**Reply by severussundar** [reply]
Hey @alawal98, yes, it is possible to filter based on multiple fields.

For the example, you have mentioned, you could do something like this.

```graphql
{
  books(
    filter: {
      and: [
        { or: [{ id: { eq: 1 } }, { id: { eq: 2 } }, { id: { eq: 3 } }] }
        {
          or: [{ genre: { eq: "Horror" } }, { genre: { eq: "Thriller" } }]
        }
      ]
    }
  ) {
    items {
      ...
    }
  }
}

```

For reference, here is the link to the `filter` operator in DAB docs: [GraphQL Filter operation docs](https://learn.microsoft.com/azure/data-api-builder/graphql#filter)

**Reply by jaykar2020** [reply]
Hi @severussundar, how does one filter on nested objects that contain string arrays? Below does not return any data (doesn't error either).

 and: [
        { locale: { eq: "en" } }
        { contentMetadata: { values: { contains: "foo" } } }
      ]

**Reply by severussundar** [reply]
Hey @jaykar2020 , the syntax you have mentioned in the comment is correct.

Only those contents (assuming `contents` as the top level entity here) items  
a) Having `locale` as `en` and 
b) `values` (field of the `contentMetadata` entity) containing `foo` in them. Here, my understanding, is that `contentMetadata` is of array type (essentially, a list of `contentMetadata` items can be associated with a single `content` item). For a given `content` item, if none of the none of the associated `contentMetadata` items have `values` items with `foo` in them, then, that particular `content` item will not come up in the response.  

It might be the case that none of the rows present in the table satisfy the conditions.

**Comment by jaykar2020** [comment]
@severussundar, thanks for the response. I've verified data is accurate. ContentMetadata is a nested object [array in the root object] and includes a string array, values. I've added the schema and sample data. The sample document should be returned as response but does not. Any idea why?

Schema
````
type ContentEntity @model{
locale: String
contentMetadata: [ContentMetadata]
}

type ContentMetadata {
values: [String]
...removed for brevity
}
````
Data in Azure Cosmos DB
````
{
    "locale": "en",
    "contentMetadata": [
        {
            "values": [ "foo" ]
        }
    ]
}
````
Edited to add: replacing string array with an array of another object (see updated schema below), works. Would really like to see the string array work though...
````
type ContentMetadata {
someObject: [SomeObject]
}

type SomeObject {
val: String
}
````
query 
````
and: [
        { locale: { eq: "en" } }
        {
          contentMetadata: {
            someObject: { val: { eq: "foo" } }
          }
        }
      ]
````
Data in Azure Cosmos DB
````
{
    "locale": "en",
    "contentMetadata": [
        {
            "someObject": [ { "val": "foo" } ]
        }
    ]
}
````

**Reply by severussundar** [reply]
Hey @jaykar2020,
 
For Cosmos DB database type, the support for filtering on nested objects is introduced in the recent 0.11.* version. 
[0.11.12-rc](https://www.nuget.org/packages/Microsoft.DataApiBuilder/0.11.12-rc) is the latest rc version in which this feature should be available. Please give it a try with this version if you were previously trying with an older version of DAB.

**Comment by jaykar2020** [comment]
@severussundar,  updated to 0.11.12-rc but still see the same issue. If this case works for you, do you mind sharing the Cosmos DB document, query, and response? I'd like to compare with mine to see if it's user error. 

![image](https://github.com/Azure/data-api-builder/assets/160181361/e44b2d7f-b784-4876-812c-1887bbbf931e)

And the data is accurate

![image](https://github.com/Azure/data-api-builder/assets/160181361/45acf3c1-6bc2-478f-b5a8-21e6db94e946)


Thanks for looking into this! It's much appreciated!

**Reply by jaykar2020** [reply]
@severussundar, any update on this? Can you confirm it's fixed in  0.11.12-rc? Thanks!
